### ISSUE TITLE
cancel should not call finish, should it?

### ISSUE THREAD

**Issue by t089** [pluralsight/PSOperations#74]
I am a bit surprised to see that the `cancel` implementation of `Operation` calls `finish` by itself, if `state > .Ready`. Shouldn't `cancel` only mark the operation as _cancelled_ ? Isn't it the responsibility of the operation to check for this flag, and if `true`, stop its work, clean up and only then transition to the `finished` state? Or am I missing something?

**Comment by jcarroll-mediafly** [comment]
Yeah, I struggled with this a bit myself. Especially when using group operations and dealing with race conditions.

If you haven't seen my reported issue I would like to hear other's thoughts on it. Shameless  plug for advice :D

https://github.com/pluralsight/PSOperations/issues/69

**Comment by MarkQSchultz** [comment]
@t089 I agree. This was pointed out quite a while ago too. One of the reasons we haven't changed it yet is that it's a pretty big breaking change.

**Comment by t089** [comment]
Ok, well I guess it is a big change, but it is also quite a big mistake in the implementation. In its current form a call to cancel will immediately fire the completion observers and clients would think that the op finished its work, while in reality it might still be very busy and do stuff. Even worse, when it is done it can't call the didFinish observer again because it already fired.

**Comment by MarkQSchultz** [comment]
You do realize that I already said that I agree with you, right? Not sure the point of most of that comment...

**Comment by t089** [comment]
Yup just wanted to add some more arguments pro change. Sorry if you feel it was redundant.

**Comment by MarkQSchultz** [comment]
👍This would be a good change. Just a big one at this point. That's all. I would want to version the framework on this change though, as it is a breaking change.

**Comment by priteshshah1983** [comment]
It doesn't look like this change was ever merged. Can you please confirm?
